File: problems/3001/solution.rb
# @return {Integer}
def min_moves_to_capture_the_queen(a, b, c, d, e, f)
  between = ->(x, y, z) { [x, z].min < y && y < [x, z].max }
  rook_blocked = (a == e && c == a && between.call(b, d, f)) ||
                 (b == f && d == b && between.call(a, c, e))
  return 1 if (a == e || b == f) && !rook_blocked

  bishop_attacks = (c - e).abs == (d - f).abs
  bishop_blocked = bishop_attacks && (a - e).abs == (b - f).abs &&
                   between.call(c, a, e) && between.call(d, b, f)
  bishop_attacks && !bishop_blocked ? 1 : 2
end

File: problems/3001/solution.erl
-spec min_moves_to_capture_the_queen(integer(), integer(), integer(), integer(), integer(), integer()) -> integer().
min_moves_to_capture_the_queen(A, B, C, D, E, F) ->
    Between = fun(X, Y, Z) -> erlang:min(X, Z) < Y andalso Y < erlang:max(X, Z) end,
    RookBlocked = (A =:= E andalso C =:= A andalso Between(B, D, F)) orelse
                  (B =:= F andalso D =:= B andalso Between(A, C, E)),
    case (A =:= E orelse B =:= F) andalso not RookBlocked of
        true -> 1;
        false ->
            BishopAttacks = abs(C - E) =:= abs(D - F),
            BishopBlocked = BishopAttacks andalso abs(A - E) =:= abs(B - F) andalso
                            Between(C, A, E) andalso Between(D, B, F),
            case BishopAttacks andalso not BishopBlocked of true -> 1; false -> 2 end
    end.
